### ISSUE TITLE
[bug] self.install_folder and self.build_folder are not available in the layout method

### ISSUE THREAD

**Issue by boussaffawalid** [conan-io/conan#9872]
Hello,

I'm trying to use the new layout method to define where generator files are being created, I tried both:
```
def layout(self):
    self.folders.generators = os.path.join(self.build_folder, "conan")
```
and

```
def layout(self):
    self.folders.generators = os.path.join(self.install_folder, "conan")
```
But in both cases `self.build_folder` and `self.install_folder` are not defined. 

How can I define self.folders.generators relatively to the build folder.
 

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows
  * Compiler+version: msvc 16
  * Conan version: 1.41.0
  * Python version: 3.7

Best,
Walid

**Comment by memsharded** [comment]
That is exactly the purpose of the layouts, to define the build folder. (Take into account that the install_folder is also going to be removed, only makes sense for some legacy). 

You need to write:
```python
def layout(self):
    self.folders.build = "mybuild"
    self.folders.generators = os.path.join(self.folders.build, "conan")
```
or something more real would be like:
```python
def layout(self):
        self.folders.build = "build-{}".format(self.settings.build_type)
        self.folders.generators = "{}/generators".format(self.folders.build)
```

In all cases, the layout is relative to the conanfile

**Comment by boussaffawalid** [comment]
We do out of source build, this means that we should be able to use any directory as a build directory.
For example: Assuming conanfile is located in: `c:\src\projects\project1\conanfile.py` and we want to build in 
`c:\build\projects\project1\build1`, how can we defined that in the layout method ?

**Comment by memsharded** [comment]
```python
def layout(self):
    self.folders.build = "build1"
    ....
```

Layouts encode a given convention over the layout. It can be any one, you can define your own logic, and implement any layout that you want. But it has to be defined in the recipe. Otherwise it is impossible for Conan to know where the built artifacts are, which was the origin of many inconsistencies, bugs and complexities in Conan.

**Comment by boussaffawalid** [comment]
Unfortunately, the solution above is not going to work, please note `c:\src\` vs  `c:\build\`.
I think I will end up using something like
```
def layout(self):
    self.folders.build = tools.get_env("MY_BUILD_FOLDER", "default_build_folder")
```

and set `MY_BUILD_FOLDER` in a helper script that calls `conan install` command

**Comment by boussaffawalid** [comment]
I dont really think that a recipe should define the build folder. Build and source folders should be completely independent. 
With the current layout approach, build folder must be relative to the source folder which not convenient.

**Comment by memsharded** [comment]
> I dont really think that a recipe should define the build folder. Build and source folders should be completely independent.
With the current layout approach, build folder must be relative to the source folder which not convenient.

We have seen too many users issues, struggles, bugs and complexities because of this. It has been this way for the last 4 years and it didn't work. Your suggestion for example makes the possibility of working on several packages at the same time (``editables``) almost impossible. It also introduces a lot of confusion among developers and make issues more difficult to debug. 
It is most likely fine if you want to define your layout in a complete different location via ``self.folders.build = tools.get_env("MY_BUILD_FOLDER", "default_build_folder")``. Note, however, that such definition can easily cause conflicts with other builds, so you should go for something like:
```python
base =  tools.get_env("MY_BUILD_FOLDER")
if base is None:
     self.folders.build = "default_build_folder"  # local to the current conanfile.py
else:  # assume it is an absolute path, it will do something like ``D:/buildfolder/pkg/version/build``
     self.folders.build = os.path.join(base, self.name, self.version, "build")
```

the important bit here is that this definition is actually encoded somewhere, otherwise, that indefinition is what makes thing worse.

**Comment by boussaffawalid** [comment]
@memsharded is there a good reason why `self.build_folder` is not available inside the `layout` method ? is it maybe possible to make it available ? 

Can we at least make `self.folders.build` defaulted to `self.build_folder`

**Comment by memsharded** [comment]
> @memsharded is there a good reason why self.build_folder is not available inside the layout method ? is it maybe possible to make it available ?

Yes. ``self.build_folder`` is computed from the ``self.folders.build``, see the implementation in ``conanfile.py``:

```python
    @property
    def build_folder(self):
        return self.folders.build_folder
````
which in turn, will be computed as the composition of the base folder + ``self.folders.build``:

```python
@property
    def build_folder(self):
        if self._base_build is None:
            return None
        if not self.build:
            return self._base_build
        return os.path.join(self._base_build, self.build)
```

In short: you cannot use ``self.build_folder`` in ``layout()`` because it is not defined yet. The main purpose of ``layout()`` is precisely to define such build folder location. Note that command line parameters like ``--build-folder`` and ``--source-folder`` will also disappear in Conan 2.0.

**Comment by boussaffawalid** [comment]
@memsharded I understand that I cannot access `self.build_folder` in the layout function as it is not defined yet. But now how can I access source folder full path in the layout method ? 
`self.source_folder` is not defined and `os.path.dirname(__file__)` points to the recipe folder!

**Comment by boussaffawalid** [comment]
ping @memsharded, is it at least possible to access source_folder **full path** in the layout method ?

**Comment by memsharded** [comment]
Hi, I think it is important to highlight that there is no global ``source_folder`` to be accessed in the ``layout()`` method. This method can only define relative layouts, because the absolute folders do change over time.

This is more or less the simplified design of the cache, for a given package reference:
- A "cache-export" folder containing the conanfile.py recipe
- A "cache-source" folder containing the sources necessary to build, for all package binaries
- For each package binary a "cache-build" folder

When a recipe is used in a user folder, with ``conan install`` + cmake or ``editable`` package, it basically uses all those folders relative to the current ``conanfile.py`` location (we will add functionality for --source-folder and --build-folder args soon)

When the recipe is in the cache, the folders keep changing:
- When the ``source()`` method execute, the global ``self.source_folder = os.path.join("cache-source", layout.relative_source)``
- When the ``build()`` method executes, as for building all the source code is copied to the "cache-build" folder, the ``self.source_folder = os.path.join("cache-build", layour.relative_source)``. What is in this "cache-build" folder mimics the local layout that the developer had in their user folder when doing "conan install".

So in short, there is not an absolute ``source_folder`` that can be accessed in the ``layout()`` method, such folder doesn't exist, it is a variable, moving thing. 

The key idea is that ``layout()`` is not an "action" method. It is not used for doing anything, it is just a declarative description of where things will be found.

I think what you are looking for is this PR: https://github.com/conan-io/conan/pull/10274, that will add ``--source-folder`` and ``--build-folder`` arguments to local flow for ``layout()`` method, could be?

**Comment by boussaffawalid** [comment]
@memsharded https://github.com/conan-io/conan/pull/10274 is exactly what I'm looking for, Thank you very much. 
I'm a bit confused however as I thought `--source-folder` and `--build-folder` are going to be deleted in future conan versions

**Comment by memsharded** [comment]
> I'm a bit confused however as I thought --source-folder and --build-folder are going to be deleted in future conan versions

You are right. We have been trying to remove them, but it seems some user feedback (including yours IIRC) was pointing to their need, so we are now considering keeping them (and providing the correct support, because one of the reasons we wanted to remove them was because they were not working well in some situations).

**Comment by boussaffawalid** [comment]
You are right, 100 voting for keeping them as well.

**Comment by memsharded** [comment]
Sorry this was not followed up back then.

Things have changed since then, Conan 2 removed ``--source-folder`` and ``--build-folder`` in favor of only ``--output-folder`` which wouldn't be recommended for most cases and instead relying on the ``layout()`` would be recommended.

Closing this issue as solved in Conan 2/outdated, please open new tickets against Conan 2 for any further question or issue, thanks for the feedback!
